src/cow-react/modules/tokens/hooks/useOnchainBalances.ts

```diff
@@ -0,0 +1,144 @@
+import { Interface } from '@ethersproject/abi'
+import { isAddress } from '@src/utils'
+import { CurrencyAmount, Token } from '@uniswap/sdk-core'
+import JSBI from 'jsbi'
+import ERC20ABI from 'abis/erc20.json'
+import { Erc20Interface } from 'abis/types/Erc20'
+import { useMultipleContractSingleData } from 'lib/hooks/multicall'
+import { ListenerOptionsWithGas } from '@uniswap/redux-multicall'
+import { useMemo } from 'react'
+import { OnchainState } from '../types'
+
+const ERC20Interface = new Interface(ERC20ABI) as Erc20Interface
+const DEFAULT_LISTENER_OPTIONS: ListenerOptionsWithGas = { gasRequired: 185_000, blocksPerFetch: 5 }
+
+export type OnchainTokenAmount = OnchainState<CurrencyAmount<Token> | undefined>
+
+export type OnchainTokenAmounts = {
+  [tokenAddress: string]: OnchainTokenAmount
+}
+
+export type OnchainAmountsResult = {
+  amounts: OnchainTokenAmounts
+  isLoading: boolean
+}
+
+export interface OnchainAmountsParams {
+  account?: string
+  tokens?: (Token | undefined)[]
+  blocksPerFetch?: number
+}
+
+export type OnchainBalancesParams = OnchainAmountsParams
+
+export type OnchainAllowancesParams = OnchainAmountsParams & { spender?: string }
+
+export function useOnchainBalances(params: OnchainBalancesParams): OnchainAmountsResult {
+  const { account } = params
+  const callParams = useMemo(() => [account], [account])
+  return useOnchainErc20Amounts('balanceOf', callParams, params)
+}
+
+export function useOnchainAllowances(params: OnchainAllowancesParams): OnchainAmountsResult {
+  const { account, spender } = params
+  const callParams = useMemo(() => [account, spender], [account, spender])
+  return useOnchainErc20Amounts('allowance', callParams, params)
+}
+
+export interface OnchainBalancesAndAllowancesParams {
+  account: string | undefined
+  spender: string | undefined
+  tokens: Token[]
+  blocksPerFetchBalance?: number
+  blocksPerFetchAllowance?: number
+}
+
+export interface OnchainBalancesAndAllowances {
+  balances: OnchainTokenAmounts
+  allowances: OnchainTokenAmounts
+  isLoading: boolean
+}
+
+/**
+ * Fetches
+ * @param params
+ * @returns
+ */
+export function useOnchainBalancesAndAllowances(
+  params: OnchainBalancesAndAllowancesParams
+): OnchainBalancesAndAllowances {
+  const { account, spender, tokens, blocksPerFetchAllowance, blocksPerFetchBalance } = params
+
+  const { amounts: balances, isLoading: areBalancesLoading } = useOnchainBalances({
```

Comment:
I would split balances and allowances, because in some cases we need only one of them.
I suggest smth like this:
```
// Updater.ts

function Updater() {
  const updateAllowancesAtom = useUpdateAtom(allowancesAtom)

  const { amounts: allowances, isLoading } = useOnchainAllowances({
    account,
    tokens,
    spender,
    blocksPerFetch: blocksPerFetchAllowance,
  })
  
  useEffect(() => {
    updateAllowancesAtom({allowances, isLoading})
  }, [allowances, isLoading])
}


// useOnchainBalancesAndAllowances.ts

function useOnchainBalancesAndAllowances() {
  const { allowances, isLoading: areAllowancesLoading } = useAtomValue(allowancesAtom)
  const { balances, isLoading: areBalancesLoading } = useAtomValue(balancesAtom)

  return useMemo(
    () => ({
      balances,
      allowances,
      isLoading: areBalancesLoading || areAllowancesLoading,
    }),
    [balances, allowances, areBalancesLoading, areAllowancesLoading]
  )
}
```

Reply:
Splitting is a good idea, but I don't understand why to add joitai in the mix. 

Both are stored in REDUX already (multi-call data). I was planning on separating (the onchain hooks are already separated, I just need the "effective" ones)